crates/hir_def/src/data.rs

```diff
@@ -105,6 +105,75 @@ impl FunctionData {
     }
 }
 
+fn expand_lazy_type_ref(
+    db: &dyn DefDatabase,
+    item_tree: &ItemTree,
+    file_id: HirFileId,
+    module_id: ModuleId,
+    lazy_type_ref: &LazyTypeRef,
+) -> Option<Interned<TypeRef>> {
+    let macro_call = match lazy_type_ref {
+        LazyTypeRef::Ty(tyref) => return Some(tyref.clone()),
+        LazyTypeRef::Unknown => return None,
+        LazyTypeRef::Macro(id) => id,
+    };
+    let macro_call = item_tree[*macro_call].clone();
+    let ast_id_map = db.ast_id_map(file_id);
+    let root = db.parse_or_expand(file_id).unwrap();
+    let macro_call = ast_id_map.get(macro_call.ast_id).to_node(&root);
+
+    let mut expander = Expander::new(db, file_id, module_id);
+    let expanded = expand(db, &mut expander, &macro_call, true)?;
+    let node = ast::Type::cast(expanded)?;
+
+    let ctx = LowerCtx::new(db, file_id);
+    return Some(Interned::new(TypeRef::from_ast(&ctx, node)));
+
+    fn expand(
```

Comment:
I've found this recursive expansion/syntax rewriting pattern all across r-a. I'd be nice if `Expander` would have a helper for this.

Reply:
Yeah.... SyntaxRewriter isn’t really intended for this in the first place...

The original idea was to *never* mutate ast in-place, and instead maintain explicit links between macro calls and expansions.